Ledger/display address and add event with PreparedTransactionData

Thoralf-M opened the pull request:
# Description of change

Display new receive and remainder address with ledger accounts so the user can verify them
Also added PreparedTransactionData event
Merge get_ledger_status and get_ledger_opened_app

## Links to any relevant issues

https://github.com/iotaledger/firefly/issues/1238

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Generated new addresses and saw the prompt on the ledger

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

Thoralf-M committed:
add get_ledger_opened_app

Thoralf-M committed:
fmt imports

Thoralf-M committed:
regenerate addresses on ledger for remainder address and get_unused_address

Thoralf-M committed:
add remainder address during syncing

Thoralf-M committed:
add PreparedTransactionData event

Thoralf-M committed:
resolve conflicts

Thoralf-M committed:
merge get_ledger_status and get_ledger_opened_app

Thoralf-M committed:
update example

Thoralf-M committed:
update LedgerStatus

Thoralf-M committed:
update LedgerApp struct

Thoralf-M committed:
update iota-ledger

Thoralf-M committed:
remove tags from ledger status

Thoralf-M committed:
update ledger dependency

Thoralf-M committed:
fix LedgerApp field names

Thoralf-M committed:
return locked status with iota app

shufps commented on src/signing/ledger.rs:
```diff
@@ -60,27 +60,24 @@ impl super::Signer for LedgerNanoSigner {
+        let (connected_, locked) =
+            match iota_ledger::get_ledger(crate::signing::ledger::HARDENED, is_simulator).map_err(Into::into) {
+                Ok(_) => (true, false),
+                Err(crate::Error::LedgerDongleLocked) => (true, true),
```
I think `get_ledger` never returns `LedgerDongleLocked`. There are commands that are allowed although the Ledger is connected. But there are commands that could be executed afterwards to figure out if it's locked. For instance the `get_data_buffer_state` call that always can be executed is only allowed if the ledger is not locked. I think, that would be a solid test.

shufps replied:
This call should return if the ledger is locked or not:

https://github.com/iotaledger/ledger.rs/commit/4d0656cedb744686fdb342ede15ce47ec92c532d#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R271

Thoralf-M committed:
return address in GeneratingRemainderDepositAddress event

Thoralf-M committed:
return object for GeneratingRemainderDepositAddress

Thoralf-M committed:
add TransactionIO struct

Thoralf-M committed:
Merge branch 'dev' into ledger/display-change-address-and-PreparedTransactionData

Thoralf-M committed:
remove GetLedgerOpenedApp

Thoralf-M commented:
@kwek20 could you please look into this `TransferProgressType` enum in java_glue.rs?
```
    /// Generating remainder value deposit address.
    GeneratingRemainderDepositAddress(AddressData),
    /// Prepared transaction.
    PreparedTransaction(PreparedTransactionData),
```
This fields hold a struct now and I didn't see how one can map it there

Thoralf-M committed:
add LedgerAddressGeneration event (#692)

* add LedgerAddressGeneration event

* update comment

Thoralf-M committed:
remove unused import

thibault-martinez commented on src/account/mod.rs:
```diff
@@ -665,7 +667,39 @@ impl AccountHandle {
+                emit_ledger_address_generation(&account, address.address().to_bech32()).await;
+
+                log::debug!("get_unused_address regenerate address so it's displayed on the ledger");
+                let regenrated_address = crate::address::get_address_with_index(
```
Typo regenrated_address

Thoralf-M committed:
fix typo

Thoralf-M committed:
try to get bech32_hrp from first account

Thoralf-M committed:
add allow_create_multiple_empty_accounts to AccountToCreate

Commit:
Ledger/display change address and prepared transaction data (#717)

* Updated TransferProgressType

* Added new TransferProgressType structs

* And added to java glue

Thoralf-M committed:
add migration address validation

Thoralf-M committed:
Ledger input spent status (#721)

* Update nodejs readme (#685)

Add LLVM dependency

* Add dust allowance outputs support (#678)

* add option to create dust_allowance_outputs

* Add option to define output_kind for transfers

* fix transfer example

* fmt

* add default_output_kind

* fmt

* Add required dependency for example (#691)

* add GetBalance to API timeouts (#667)

* remove "$" from codeblocks and update nodejs getting started (#693)

* Migrate to docusaurus (#668)

* Removed mdbooks and migrated to docusarus.
Fixed broken links.
Moved images to docs/static/.

* added favicon, deleted old config, fixed rust api ref

* * edited welcome section
* added package-lock.json

* * Edited Overview and Libraries Overview.
* Edited Libraries/Rust

* * Edited Libraries/Nodejs

* * Edited Python landing and troubleshooting.md.
* Moved image to static/libraries/python/

* * Edited Python examples.md and api_reference.md
* Moved image to static

* * Edited nodejs/api_reference.md
* Edited python/api_reference.md
* Edited contribute.md
* Edited python/examples.md
* Edited Specs/README.md

* * Updated favicon

* * Re-added image which shouldn't've been deleted

* * Formatted titles

* * Reviewed imgs alt texts

* * Using uppercase always for Stronghold and Wallet.rs

* * Changing use of `` for italics in most cases except IOTA products such as Wallet.rs, Stronghold, Chrysalis, Firefly

* * Changing use of `` for italics in most cases except IOTA products such as Wallet.rs, Stronghold, Chrysalis, Firefly

* Updated editUrl

* * Corrected package npmjs url and status

* * Rephrased introduction to libraries

* * Rephrased introduction to libraries

* *Deleted useless files.
* Deleted useless readme section
* Correct Python lib readme  title

* * Renaming parent documentation dir to docOps to avoid the docs/docs path as per @huhn511 requested

* * Updating node/api_reference with latest changes.
* Addressing Thoralf's comments.
* Updated dependencies to use the `main` branch

* Update docOps/docs/welcome.md

Co-authored-by: Aleksei Korolev <nice.guy.alex.korolev@gmail.com>

* Update docOps/docs/welcome.md

Co-authored-by: Aleksei Korolev <nice.guy.alex.korolev@gmail.com>

* Update docOps/docs/libraries/python/examples.md

Co-authored-by: Aleksei Korolev <nice.guy.alex.korolev@gmail.com>

* Update docOps/docs/libraries/README.md

Co-authored-by: Aleksei Korolev <nice.guy.alex.korolev@gmail.com>

* Fixing title in libraries/README.md

* * Replaced <code></code> and <strong></strong> tags with MD syntax in docOps folder

* * Removed HTML tables and tags.
* Fixed broken links

* * Changed iota-wallet git

* * Added EOF emtpy LF

* * Merged intro into one paragraph

* * Added NodeInfo table

* * Added clarifications about test_net and tangle explorer in libraries/README.md
* Formatted specs/README.md

* * Removed security warning from nodejs and python examples sections as it was already in the getting started section

* * Addressed PR comments by @aleksei-korolev

* * Formatted returns in python/api_reference.md
* Added Rust code support
* Formatted nodejs/examples.md
* Added docs links for dependencies in libraries/rust/getting_started.md
* Fixed punctuation in libraries/README.md
* Added line break in table in specs/README.md

* add custom IOTA design

* add new gh action

* fix folder name

* fix baseUrl

* Update docOps/docs/libraries/nodejs/examples.md

Co-authored-by: Aleksei Korolev <nice.guy.alex.korolev@gmail.com>

* Update docOps/docs/libraries/nodejs/examples.md

Co-authored-by: Aleksei Korolev <nice.guy.alex.korolev@gmail.com>

* Update docOps/docs/libraries/python/examples.md

Co-authored-by: Aleksei Korolev <nice.guy.alex.korolev@gmail.com>

* * Removed <ul> and <li> elements from .md tables
* Fixed blank lines
* Changed function emphasis from __ to ``
* Removed contractions
* Added references

* * Updated theme
* Renamed folder README files to index
* Added nav and sidebar

* * Updated routes

* * Fixed footer links
* Fixed logo and favicon paths

* * Attempting to trigger pipeline

* add rust deployment script

* Install rust-docs with nightly toolchain

* * Updated cname and trigger branches on deploy_docs_to_gh-pages.yml
* Set dark theme as default and updated style sheet
* Updated landing page to use config from docusaurus.config.js for title and slug
* Updated landing page action texts

* * Removed branch from deploy_docs_to_gh-pages.yml
* Removed /wallet.rs/ from base url in docusaurus.config.js

* * [REVERT] - Removed branch from deploy_docs_to_gh-pages.yml
* [REVERT] - Removed /wallet.rs/ from base url in docusaurus.config.js
* Added missing svg and stylesheet

* * Removed deploy rust docs action from deploy_docs_to_gh-pages.yml

* * Fixed landing page links

* * Removed current branch from deployable branches
* Removed /wallet.rs/ from baseUrl config

* * Removed watermark from static/img/specs/erd.jpg

* * Removed language specific overview and troubleshooting pages
* Added general troubleshooting section
* Forcing libraries sidebar to be open
* Added docusaurus-plugin-includes package
* Readded include statements
* Modified config to force deploy at local GH pages

* * Reverted - Modified config to force deploy at local GH pages

* * Replaced erd.jpg with erdIOTA.svg

* * Editing config to force deploy

* * Reverted - Editing config to force deploy

* * Renamed deploy_docs_to_gh-pages.yml to deploy-docs-to-gh-pages.yml
* Renamed docOps directory to documentation
* Updated deploy-docs-to-gh-pages.yml script to use documentation directory

Co-authored-by: Lucas Tortora <lucastortora@gmail.com>
Co-authored-by: Aleksei Korolev <nice.guy.alex.korolev@gmail.com>
Co-authored-by: huhn511 <sehe89@gmail.com>

* chore(deps): bump prismjs from 1.23.0 to 1.24.1 in /documentation (#695)

Bumps [prismjs](https://github.com/PrismJS/prism) from 1.23.0 to 1.24.1.
- [Release notes](https://github.com/PrismJS/prism/releases)
- [Changelog](https://github.com/PrismJS/prism/blob/master/CHANGELOG.md)
- [Commits](https://github.com/PrismJS/prism/compare/v1.23.0...v1.24.1)

---
updated-dependencies:
- dependency-name: prismjs
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* Update README.md

* Start and stop background sync messages (#698)

* add start and stop background sync messages to the actor

* fix nodejs binding

* update bindings and versions

* Don't return error for spent outputs (#699)

* Ledger/display change address and prepared transaction data (#700)

* add get_ledger_opened_app

* fmt imports

* regenerate addresses on ledger for remainder address and get_unused_address

* add remainder address during syncing

* add PreparedTransactionData event

* resolve conflicts

* merge get_ledger_status and get_ledger_opened_app

* update example

* update LedgerStatus

* update LedgerApp struct

* update iota-ledger

* remove tags from ledger status

* update ledger dependency

* fix LedgerApp field names

* return locked status with iota app

* return address in GeneratingRemainderDepositAddress event

* return object for GeneratingRemainderDepositAddress

* add TransactionIO struct

* remove GetLedgerOpenedApp

* add LedgerAddressGeneration event (#692)

* add LedgerAddressGeneration event

* update comment

* Don't return error for spent outputs (#694)

* return spent status for inputs

* Ignore spent output (#703)

* Don't return error for spent outputs

* fmt

* Don't return error for spent outputs (#694)

* don't return error for spent output

* Revert "Merge branch 'firefly-testing' into ledger-input-spent-status"

This reverts commit 44f18f5722203a96ced37923466afa669aa3f079, reversing
changes made to 029d01b8962e5bc56eeca84f9c9811a2631fb31b.

* fix spent_addresses for no inputs

* add spent_addresses to MigrationMetadata

Co-authored-by: lucas-tortora <85233773+lucas-tortora@users.noreply.github.com>
Co-authored-by: Lucas Tortora <lucastortora@gmail.com>
Co-authored-by: Aleksei Korolev <nice.guy.alex.korolev@gmail.com>
Co-authored-by: huhn511 <sehe89@gmail.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Thoralf-M committed:
add account_id to migration address function (#715)

Thoralf-M committed:
Merge branch 'dev' into ledger/display-change-address-and-PreparedTransactionData

Thoralf-M committed:
Merge branch 'dev' into ledger/display-change-address-and-PreparedTransactionData

Thoralf-M committed:
fix variant index

Thoralf-M committed:
remove GetLedgerOpenedApp

melatron commented on src/signing/stronghold.rs:
```diff
@@ -28,12 +28,12 @@ pub(crate) async fn stronghold_path(storage_path: &Path) -> crate::Result<PathBu
-        Err(crate::Error::NoLedgerSignerError)
+        // dummy status
+        crate::LedgerStatus {
+            connected: false,
```
shouldn't the default dummy status be connected: true?

Thoralf-M replied:
I would rather say no, because it's not a ledger, but at the end it shouldn't be used at all, it's only required for the trait

melatron commented on src/event.rs:
```diff
@@ -702,6 +768,40 @@ pub(crate) async fn emit_address_consolidation_needed(account: &Account, address
+#[cfg(any(feature = "ledger-nano", feature = "ledger-nano-simulator"))]
+#[cfg_attr(docsrs, doc(cfg(any(feature = "ledger-nano", feature = "ledger-nano-simulator"))))]
+pub async fn on_ledger_address_generation<F: Fn(&LedgerAddressGeneration) + Send + 'static>(cb: F) -> EventId {
+    let mut l = ledger_address_generation_listeners().lock().await;
```
It would be more consistent to name `l` as `listeners`.

melatron commented on src/event.rs:
```diff
@@ -90,6 +90,18 @@ pub struct AddressConsolidationNeeded {
+    #[serde(rename = "accountId")]
+    /// The associated account identifier.
+    pub account_id: String,
+    /// The transfer event type.
```
is `addressData` the event type? The naming is kind of misleading

Thoralf-M replied:
The Event type is LedgerAddressGeneration or TransferProgressType, I will update to comment for the struct

Thoralf-M committed:
Add clearer comments and use full name for variable

Thoralf-M committed:
update node for tests
